### PR TITLE
Use controller runtime for federated seed controllers

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/area dev-productivity
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR makes the federated seed controller (extensions workers) use the [Controller-Runtime](https://github.com/kubernetes-sigs/controller-runtime) manager and controllers in favor of generated informers/listers. It's a further and consequent step towards the Controller-Runtime and allows to get rid of any third party generated client code (e.g. from external-dns-management) which turned out to be a blocker for #2489.

**Special notes for your reviewer**:
I tested several scenarios, e.g. shoot creation/deletion, adding/removing workers with "new" OS type, seed creation/deletion, provoking errors in the extension controllers, etc.. It'd be nice if you could double check the state reconciliation as well (/cc @plkokanov @swilen-iwanow).

The controller manager that is built for each seed in the Gardenlet now uses the client/informer cache introduced by @tim-ebert.

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

**Comment by timuthy** [comment]
Update:
- Resolved golang-ci issues
- File renaming
- Small channel improvements

**Review by timebertt — COMMENTED** [review]
Looks quite good!

I haven't looked into it in detail, but I am slightly worried about how that will play together with invalidating the seed client in the client map. I.e. if the Seed's kubernetes version is upgraded, the client will be invalidated automatically, which means that all the informers/watches will be stopped. That in turn means, that the controllers you have here won't receive any watch events anymore.
Have you tested this scenario?

**Comment by timuthy** [comment]
> I haven't looked into it in detail, but I am slightly worried about how that will play together with invalidating the seed client in the client map. I.e. if the Seed's kubernetes version is upgraded, the client will be invalidated automatically, which means that all the informers/watches will be stopped. That in turn means, that the controllers you have here won't receive any watch events anymore.

Nice scenario but no, I didn't test this. Let me check again and come back. But it will probably take me some days since I'm out of office next week.

**Comment by plkokanov** [comment]
> Looks quite good!
> 
> I haven't looked into it in detail, but I am slightly worried about how that will play together with invalidating the seed client in the client map. I.e. if the Seed's kubernetes version is upgraded, the client will be invalidated automatically, which means that all the informers/watches will be stopped. That in turn means, that the controllers you have here won't receive any watch events anymore.
> Have you tested this scenario?

Not sure if the gardenlet leader election mechanism uses the same clients. If so invalidating the client would cause the federated controllers to shut down and gardenlet to restart.
So to cover the cases where one gardenlet manages multiple seeds maybe we should implement a federated leader-election mechanism for the federated controllers?

**Comment by timebertt** [comment]
> Not sure if the gardenlet leader election mechanism uses the same clients. If so invalidating the client would cause the federated controllers to shut down and gardenlet to restart.

Nope, the leader election is using its own client:

https://github.com/gardener/gardener/blob/398870a9a2db8eb5eaf4489538a64f20ff6cd7fd/cmd/gardenlet/app/gardenlet.go#L316-L323

> So to cover the cases where one gardenlet manages multiple seeds maybe we should implement a federated leader-election mechanism for the federated controllers?

Hmm, I'm not sure what you mean by "federated leader-election". Can you elaborate?
(Anyhow, using the gardenlet for multiple seeds is not recommended and mostly only a development scenario)

**Comment by plkokanov** [comment]
Currently (afaik) there is no way to stop a federated controller's control loop  from running for a particular seed (unless the seed is deleted). So in the case of one gardenlet managing multiple seeds you could have a scenario where a particular seed should no longer be managed by the same gardenlet, but by another. You would want to have a mechanism which can stop the federated controller for that seed and start it in the new gardenlet.

But as you mentioned, this is not a recommended setup and will most likely be encountered only in a development scenario. Also I believe we're side tracking a bit from the PR :).

Anyway, the PR looks good. 
I'll test it out as well a bit later today.

**Comment by plkokanov** [comment]
Ran the PR locally but the `Infrastructure` and `Worker` states are not synced to the `ShootState`

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.BackupEntryResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.BackupEntryResource, func() runtime.Object { return &extensionsv1alpha1.BackupEntryList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.BackupEntryResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.BackupEntryResource, func() runtime.Object { return &extensionsv1alpha1.BackupEntryList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:117]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.BackupEntryResource, func() runtime.Object { return &extensionsv1alpha1.BackupEntry{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.ContainerRuntimeResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.ContainerRuntimeResource, func() runtime.Object { return &extensionsv1alpha1.ContainerRuntimeList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.ContainerRuntimeResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ContainerRuntimeResource, func() runtime.Object { return &extensionsv1alpha1.ContainerRuntimeList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:139]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ContainerRuntimeResource, func() runtime.Object { return &extensionsv1alpha1.ContainerRuntime{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.ControlPlaneResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.ControlPlaneResource, func() runtime.Object { return &extensionsv1alpha1.ControlPlaneList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.ControlPlaneResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ControlPlaneResource, func() runtime.Object { return &extensionsv1alpha1.ControlPlaneList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:161]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ControlPlaneResource, func() runtime.Object { return &extensionsv1alpha1.ControlPlane{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.ExtensionResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.ExtensionResource, func() runtime.Object { return &extensionsv1alpha1.ExtensionList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.ExtensionResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ExtensionResource, func() runtime.Object { return &extensionsv1alpha1.ExtensionList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:183]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.ExtensionResource, func() runtime.Object { return &extensionsv1alpha1.Extension{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.InfrastructureResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.InfrastructureResource, func() runtime.Object { return &extensionsv1alpha1.InfrastructureList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.InfrastructureResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.InfrastructureResource, func() runtime.Object { return &extensionsv1alpha1.InfrastructureList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:205]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.InfrastructureResource, func() runtime.Object { return &extensionsv1alpha1.Infrastructure{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.NetworkResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.NetworkResource, func() runtime.Object { return &extensionsv1alpha1.NetworkList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.NetworkResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.NetworkResource, func() runtime.Object { return &extensionsv1alpha1.NetworkList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:227]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.NetworkResource, func() runtime.Object { return &extensionsv1alpha1.Network{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.OperatingSystemConfigResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.OperatingSystemConfigResource, func() runtime.Object { return &extensionsv1alpha1.OperatingSystemConfigList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.OperatingSystemConfigResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.OperatingSystemConfigResource, func() runtime.Object { return &extensionsv1alpha1.OperatingSystemConfigList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:249]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.OperatingSystemConfigResource, func() runtime.Object { return &extensionsv1alpha1.OperatingSystemConfig{} }),
```

File: pkg/gardenlet/controller/federatedseed/extensions/add.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+			{
+				name:                fmt.Sprintf("%s-%s", controllerInstallationName, extensionsv1alpha1.WorkerResource),
+				concurrentReconcile: controllerInstallationWorkers,
+				reconciler:          control.createExtensionRequiredReconcileFunc(ctx, extensionsv1alpha1.WorkerResource, func() runtime.Object { return &extensionsv1alpha1.WorkerList{} }),
+				updateFunc:          extensionTypeChanged,
+			},
+			{
+				name:                fmt.Sprintf("%s-%s", shootStateName, extensionsv1alpha1.WorkerResource),
+				concurrentReconcile: shootStateWorkers,
+				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.WorkerResource, func() runtime.Object { return &extensionsv1alpha1.WorkerList{} }),
```
Comment (plkokanov) [inline pkg/gardenlet/controller/federatedseed/extensions/add.go:271]:
```suggestion
				reconciler:          stateControl.createShootStateSyncReconcileFunc(ctx, extensionsv1alpha1.WorkerResource, func() runtime.Object { return &extensionsv1alpha1.Worker{} }),
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
This should fix the issue that I mentioned about syncing the state to the `ShootState`

**Comment by timuthy** [comment]
> I haven't looked into it in detail, but I am slightly worried about how that will play together with invalidating the seed client in the client map. I.e. if the Seed's kubernetes version is upgraded, the client will be invalidated automatically, which means that all the informers/watches will be stopped.

I just talked to @tim-ebert and he's going to address this case in a separate PR because it needs to be handled more generically.

@plkokanov many thanks for testing and reviewing this PR. I'll address your comments asap.

**Comment by timuthy** [comment]
PR is closed in favor of #2588 which uses the cached informers directly.

**Pull request closed without merging**